In [4]:
import numpy as np
import matplotlib.pyplot as plt
from bs4 import BeautifulSoup # For HTML parsing
import requests
import re # Regular expressions
from time import sleep # To prevent overwhelming the server between connections
from collections import Counter # Keep track of our term counts
#from nltk.corpus import stopwords # Filter out stopwords, such as 'the', 'or', 'and'
import nltk
import pandas as pd # For converting results to a dataframe and bar chart plots
%matplotlib inline

In [5]:
import csv
import datetime
import time

In [63]:
import sqlalchemy
from sqlalchemy import create_engine

In [ ]:
%load_ext watermark

# This notebook is written by Yishin and Chi-Hung.

In [118]:
%watermark

2017-01-31T19:46:42+08:00

CPython 3.5.2
IPython 5.1.0

compiler   : GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)
system     : Darwin
release    : 15.6.0
machine    : x86_64
processor  : i386
CPU cores  : 4
interpreter: 64bit


---

## First of all, we know that there are 7 types of vacuums on Amazon

In [7]:
def getVacuumTypeUrl(vacuumType,pageNum=1):
    vcleaners={"central":11333709011,"canister":510108,"handheld":510114,"robotic":3743561,"stick":510112,"upright":510110,"wetdry":553022}
    url_type_base="https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_"+str(pageNum)+"?ie=UTF8&node="
    url=url_type_base+str(vacuumType)+"&page="+str(pageNum)
    print (url)
    return url

In [8]:
vcleaners={"central":11333709011,"canister":510108,"handheld":510114,"robotic":3743561,"stick":510112,"upright":510110,"wetdry":553022}

for key in vcleaners:
    print(key,vcleaners[key])
    getVacuumTypeUrl(vcleaners[key])

canister 510108
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=510108&page=1
robotic 3743561
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=3743561&page=1
upright 510110
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=510110&page=1
central 11333709011
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=11333709011&page=1
stick 510112
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=510112&page=1
wetdry 553022
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=553022&page=1
handheld 510114
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=510114&page=1


## The following are two functions which we aim to obtain the total number of pages of each vacuum type

In [14]:
def getFinalPageNum(url,maxretrytime=20):
    passed=False
    cnt=0
    
    while(passed==False):
        cnt+=1
        print("iteration from getFinalPageNum=",cnt)
        if(cnt>maxretrytime):
            raise Exception("Error from getFinalPageNum(url)! Tried too many times but we are still blocked by Amazon.")
        try:
            with requests.Session() as session:
                session.headers = {'User-Agent': "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:50.0) Gecko/20100101 Firefox/50.0"}            
                r=session.get(url)
                if (r.status_code==200):
                    soup=BeautifulSoup(r.content,"lxml")
                    if("Robot Check" in soup.text):
                        print("we are blocked!")
                    else:
                        tagsFinalPageNum=soup.select("span[class='pagnDisabled']")
                        finalPageNum=str(tagsFinalPageNum[0].text)
                        passed=True

                else:
                    print("Connection failed. Reconnecting...")
        except:
            print("Error from getFinalPageNum(url)! Probably due to connection time out")
    return finalPageNum 

In [15]:
def InferFinalPageNum(vacuumType,pageNum=1,times=10):
    url=getVacuumTypeUrl(vacuumType,pageNum)
    
    list_finalpageNum=[]
    
    for j in range(times):
        finalpageNum=getFinalPageNum(url)
        list_finalpageNum.append(finalpageNum)
    FinalpageNum=min(list_finalpageNum)

    return FinalpageNum

In [17]:
FinalPageNum=InferFinalPageNum(510114,pageNum=1)
print('FinalPageNum=',FinalPageNum)

https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=510114&page=1
iteration from getFinalPageNum= 1
iteration from getFinalPageNum= 1
iteration from getFinalPageNum= 1
iteration from getFinalPageNum= 1
iteration from getFinalPageNum= 1
iteration from getFinalPageNum= 1
iteration from getFinalPageNum= 1
iteration from getFinalPageNum= 1
iteration from getFinalPageNum= 1
iteration from getFinalPageNum= 1
FinalPageNum= 82


So, right now, we are able to infer the total number of pages of a specific vacuum type.

## The next step is to generate all URLs of the selected vacuum type:

In [19]:
def urlsGenerator(typenode,FinalPageNum):
    #Note: 'typenode' and 'FinalpageNum' are both string

    URLs=[]
    pageIdx=1
    while(pageIdx<=int(FinalPageNum)):
        url_Type="https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_"+str(pageIdx)+"?ie=UTF8&node="
        url=url_Type+str(typenode)+"&page="+str(pageIdx)
        #print(url)
        URLs.append(url)
        pageIdx+=1
   
    return URLs

For the moment, let us choose the vacuum type "handheld":

In [20]:
URLs=urlsGenerator(510114,FinalPageNum)
len(URLs)
for url in URLs:
    print(url)

https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=510114&page=1
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_2?ie=UTF8&node=510114&page=2
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_3?ie=UTF8&node=510114&page=3
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_4?ie=UTF8&node=510114&page=4
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_5?ie=UTF8&node=510114&page=5
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_6?ie=UTF8&node=510114&page=6
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_7?ie=UTF8&node=510114&page=7
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_8?ie=UTF8&node=510114&page=8
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_9?ie=UTF8&node=510114&page=9
https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_10?ie=UTF

## Next, we'd like to obtain all the "soups" of the vacuum type "handheld" and store them into a list

In [21]:
def soupGenerator(URLs,maxretrytime=20):    

    soups=[]
    urlindex=0
    for URL in URLs:
        urlindex+=1
        print("urlindex=",urlindex)
        passed=False
        cnt=0    
        while(passed==False):
            cnt+=1
            print("iteration=",cnt)
            if(cnt>maxretrytime):
                raise Exception("Error from soupGenerator(url,maxretrytime=20)! Tried too many times but we are still blocked by Amazon.")
        
            try:
                with requests.Session() as session:
            
                    session.headers = {'User-Agent': "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:50.0) Gecko/20100101 Firefox/50.0"}            
                    r=session.get(URL)            
                
                    if (r.status_code==200):                
                        soup=BeautifulSoup(r.content,"lxml")
                        if("Robot Check" in soup.text):
                            print("we are blocked!")
                        else:
                            print("we are not blocked!")
                            soups.append(soup)
                            passed=True
                        
                    else:
                        print ("Connection failed. Reconnecting...")
            except:
                print("Error from soupGenerator(URLs,maxretrytime=20)! Probably due to connection time out")
                
    return soups 

In [10]:
soups=soupGenerator(URLs,maxretrytime=20)

urlindex= 1
iteration= 1
we are not blocked!
urlindex= 2
iteration= 1
we are not blocked!
urlindex= 3
iteration= 1
we are not blocked!
urlindex= 4
iteration= 1
we are not blocked!
urlindex= 5
iteration= 1
we are not blocked!
urlindex= 6
iteration= 1
we are not blocked!
urlindex= 7
iteration= 1
we are not blocked!
urlindex= 8
iteration= 1
we are not blocked!
urlindex= 9
iteration= 1
we are not blocked!
urlindex= 10
iteration= 1
we are not blocked!
urlindex= 11
iteration= 1
we are not blocked!
urlindex= 12
iteration= 1
we are not blocked!
urlindex= 13
iteration= 1
we are not blocked!
urlindex= 14
iteration= 1
we are not blocked!
urlindex= 15
iteration= 1
we are not blocked!
urlindex= 16
iteration= 1
we are not blocked!
urlindex= 17
iteration= 1
we are not blocked!
urlindex= 18
iteration= 1
we are not blocked!
urlindex= 19
iteration= 1
we are not blocked!
urlindex= 20
iteration= 1
we are not blocked!
urlindex= 21
iteration= 1
we are not blocked!
urlindex= 22
iteration= 1
we are not blocke

How many soups have we created?

In [11]:
print(len(soups))

82


---

## Let us pause for a while. We would like to review the usage of CSS selectors

In [181]:
example='''
<span class="abc">
  <div>
    <a href="http://123xyz.com"></a>
    hello_div01
  </div>
</span>

<span class="def">
  <a href="http://www.go.123xyz"></a>
  <div>hello_div02</div>
</span>
'''

In [182]:
mysoup=BeautifulSoup(example,"lxml")

In [183]:
print(mysoup.prettify())

<html>
 <body>
  <span class="abc">
   <div>
    <a href="http://123xyz.com">
    </a>
    hello_div01
   </div>
  </span>
  <span class="def">
   <a href="http://www.go.123xyz">
   </a>
   <div>
    hello_div02
   </div>
  </span>
 </body>
</html>


#### Exercise: look for a specific tag which is a descendent of some other tag

In [187]:
mysoup.select(".abc a")

[<a href="http://123xyz.com"></a>]

In [288]:
mysoup.select(".abc > a")

[]

the symbol > indicates that we'd like to look for a tags, which are **direct descendents** of the tag which its class=abc.

If we use ".abc a", it means that we would like to find **all descendents** of the tag which its class=abc.

In [289]:
mysoup.select(".abc > div")

[<div>
 <a href="http://123xyz.com"></a>
     hello_div01
   </div>]

#### Exercise: we look for the tags whose value of the attr href starts with "http"

In [290]:
mysoup.select("a[href^='http']")

[<a href="http://123xyz.com"></a>, <a href="http://www.go.123xyz"></a>]

#### Exercise: we look for the tags whose value of the attr href ends with "http"

In [291]:
mysoup.select("a[href$='http']")

[]

#### Exercise: extract the value of a specific attr of a specific tag

In [185]:
mysoup.select(".abc a")[0]["href"]

'http://123xyz.com'

#### more info about CSS selectors:
    
https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors

http://wiki.jikexueyuan.com/project/python-crawler-guide/beautiful-soup.html

---

In [581]:
sp=soups[70].select('li[id^="result_"]')[0]

print(sp)

for s in sp:
    try:
        print(sp.span)
    except:
        print("error")

<li class="s-result-item celwidget " data-asin="B00QBJ0NP6" id="result_1680"><div class="s-item-container"><div class="a-row sx-badge-region"><div class="a-row a-spacing-top-micro a-spacing-micro"><div class="a-row a-spacing-large"></div></div></div><div class="a-row a-spacing-base"><div aria-hidden="true" class="a-column a-span12 a-text-left"><div class="a-section a-spacing-none a-inline-block s-position-relative"><a class="a-link-normal a-text-normal" href="https://www.amazon.com/Mini-Vacuum-Cleaner-Household-Vacuuming/dp/B00QBJ0NP6/ref=lp_510114_1_1681/159-6918950-0010931?s=vacuums&amp;ie=UTF8&amp;qid=1485362182&amp;sr=1-1681"><img alt="Product Details" class="s-access-image cfMarker" data-search-image-load="" height="160" src="https://images-na.ssl-images-amazon.com/images/I/31z0+kMLpUL._AC_US160_.jpg" srcset="https://images-na.ssl-images-amazon.com/images/I/31z0+kMLpUL._AC_US160_.jpg 1x, https://images-na.ssl-images-amazon.com/images/I/31z0+kMLpUL._AC_US240_QL65_.jpg 1.5x, https:/

## Let's go back. 

### First of all, let us look for the Product URL of the first item of the first page

print the link of the first page:

In [281]:
URLs=urlsGenerator(510114,FinalPageNum)
len(URLs)
print(URLs[0])
#for url in URLs:
#    print(url)

https://www.amazon.com/home-garden-kitchen-furniture-bedding/b/ref=sr_pg_1?ie=UTF8&node=510114&page=1


We found that the Product URL of the first item can be extracted via:

In [593]:
soups[0].select('li[id^="result_"]')[0].select("a[class='a-link-normal s-access-detail-page a-text-normal']")[0]

<a class="a-link-normal s-access-detail-page a-text-normal" href="https://www.amazon.com/BLACK-DECKER-CHV1410L-Cordless-Lithium/dp/B006LXOJC0/ref=lp_510114_1_1/157-7476471-7904367?s=vacuums&amp;ie=UTF8&amp;qid=1485361951&amp;sr=1-1" title="BLACK + DECKER CHV1410L 16V Cordless Lithium Hand Vac"><h2 class="a-size-base a-color-null s-inline s-access-title color-variation-title-replacement a-text-normal" data-attribute="BLACK + DECKER CHV1410L 16V Cordless Lithium Hand Vac" data-max-rows="0" data-truncate-by-character="false">BLACK + DECKER CHV1410L 16V Cordless Lithium Hand Vac</h2></a>

where we have used the fact that each item has one unique id.

### Now, we have another goal: obtain the total number of customer reviews of the selected item (first item in the first page). Doing so we are also able to obtain the link of that item, which is pretty nice, since the item name and the item ID can be extracted from that link.

In [265]:
csrev_tag=soups[0].select('li[id^="result_"]')[0].select("a[href$='customerReviews']")[0]
print(csrev_tag)

<a class="a-size-small a-link-normal a-text-normal" href="https://www.amazon.com/BLACK-DECKER-CHV1410L-Cordless-Lithium/dp/B006LXOJC0/ref=lp_510114_1_1/157-7476471-7904367?s=vacuums&amp;ie=UTF8&amp;qid=1485361951&amp;sr=1-1#customerReviews">10,106</a>


This means we are able to obtain the total number of customer reviews (10,106) and also the link of the selected item:

https://www.amazon.com/BLACK-DECKER-CHV1410L-Cordless-Lithium/dp/B006LXOJC0/ref=lp_510114_1_1/157-7476471-7904367?s=vacuums&amp;ie=UTF8&amp;qid=1485361951&amp;sr=1-1


The above link will then be replaced by the following one:

https://www.amazon.com/BLACK-DECKER-CHV1410L-Cordless-Lithium/product-reviews/B006LXOJC0/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000

which shows 50 customer reviews per page (instead of 10 reviews per page by default).

### Another Goal: We'd like to obtain the price of the selected item

Now, let's look for more information, e.g. the price of the selected product. We know that the tag we have found is stored at the end part of a big tag which contains all the info of a specific item. Now, to retrieve more info of that item, we'll move ourselves from the end part to the front gradually.

In [266]:
csrev_tag.parent

<div class="a-row a-spacing-none"><span name="B006LXOJC0">
<span class="a-declarative" data-a-popover='{"max-width":"700","closeButton":"false","position":"triggerBottom","url":"/review/widgets/average-customer-review/popover/ref=acr_search__popover?ie=UTF8&amp;asin=B006LXOJC0&amp;contextId=search&amp;ref=acr_search__popover"}' data-action="a-popover"><a class="a-popover-trigger a-declarative" href="javascript:void(0)"><i class="a-icon a-icon-star a-star-4-5"><span class="a-icon-alt">4.3 out of 5 stars</span></i><i class="a-icon a-icon-popover"></i></a></span></span>
<a class="a-size-small a-link-normal a-text-normal" href="https://www.amazon.com/BLACK-DECKER-CHV1410L-Cordless-Lithium/dp/B006LXOJC0/ref=lp_510114_1_1/157-7476471-7904367?s=vacuums&amp;ie=UTF8&amp;qid=1485361951&amp;sr=1-1#customerReviews">10,106</a></div>

In [317]:
csrev_tag.parent.previous_sibling.previous_sibling

<div class="a-row a-spacing-mini"><div class="a-row a-spacing-none"><a class="a-link-normal a-text-normal" href="https://www.amazon.com/BLACK-DECKER-CHV1410L-Cordless-Lithium/dp/B006LXOJC0/ref=lp_510114_1_1/157-7476471-7904367?s=vacuums&amp;ie=UTF8&amp;qid=1485361951&amp;sr=1-1"><span class="a-color-base"><span class="sx-price sx-price-large">
<sup class="sx-price-currency">$</sup>
<span class="sx-price-whole">59</span>
<sup class="sx-price-fractional">99</sup>
</span>
</span></a><span class="a-letter-space"></span><span aria-label="Suggested Retail Price: $89.99" class="a-size-base-plus a-color-secondary a-text-strike">$89.99</span><span class="a-letter-space"></span><i aria-label="Prime" class="a-icon a-icon-prime a-icon-small s-align-text-bottom"><span class="a-icon-alt">Prime</span></i></div><div class="a-row a-spacing-none"><div class="a-row a-spacing-none"><span class="a-size-small a-color-secondary">Get it by <span class="a-color-success a-text-bold">Tomorrow, Jan 26</span></spa

In [326]:
pricetag=csrev_tag.parent.previous_sibling.previous_sibling
price=pricetag.select(".sx-price-whole")[0].text
fraction_price=pricetag.select(".sx-price-fractional")[0].text
print(price,fraction_price)
print(int(price)+0.01*int(fraction_price))

59 99
59.99


so, we are able to obtain the price of the selected item.

### Yet Another Goal: Let's see if we can obtain the brand of the selected item

In [333]:
pricetag.parent

<div class="s-item-container"><div class="a-row sx-badge-region"><div class="a-row a-spacing-top-micro a-spacing-micro"><a class="a-size-small a-link-normal a-text-normal" href="/gp/bestsellers/home-garden/510114/ref=sr_bs_1_510114_1"><span class="aok-float-left sx-badge-rectangle sx-bestseller-color"><span class="sx-badge-text s-color-white">Best Seller</span></span><span class="aok-float-left sx-badge-triangle sx-bestseller-color"></span><span class="s-padding-left-mini s-hidden aok-relative sx-top-left-badge-tooltip sx-bestseller-node sx-bestseller-color">in Handheld Vacuums</span></a></div></div><div class="a-row a-spacing-base"><div aria-hidden="true" class="a-column a-span12 a-text-left"><div class="a-section a-spacing-none a-inline-block s-position-relative"><a class="a-link-normal a-text-normal" href="https://www.amazon.com/BLACK-DECKER-CHV1410L-Cordless-Lithium/dp/B006LXOJC0/ref=lp_510114_1_1/157-7476471-7904367?s=vacuums&amp;ie=UTF8&amp;qid=1485361951&amp;sr=1-1"><img alt="Pr

In [335]:
pricetag.previous_sibling.parent.select(".a-size-small")[2].text

'BLACK+DECKER'

### Another goal: number of the average stars of the selected item

In [657]:
for j in range(30):
    starSelect=soups[j].select('li[id^="result_"]')[0].select_one("span[class='a-declarative']")

    if(starSelect is None):           # there are no reviews yet (hence, we see no stars at all)
        item_avestar=0
    else:
        try:
            item_avestar=starSelect.span.string.split(" ")[0]
        except:
            print(starSelect.span)
            #print(soups[j].select('li[id^="result_"]')[0])
    print(j,item_avestar)

0 4.3
1 3.9
2 4.1
3 3.8
4 3.9
5 1.8
6 3.7
7 4.3
8 0
9 4.1
10 5
11 4.3
12 0
13 4.6
14 3.6
15 2.2
16 0
17 3
18 0
19 5
20 0
21 0
22 0
None
23 0
24 5
25 0
26 0
27 1
28 0
29 3


In [646]:
for j in range(15):
    try:
        print(soups[j].select('li[id^="result_"]')[0].select_one("span[class='a-declarative']").span.string.split(" ")[0])
    except:
        print("index= ",j,", 0 stars (no reviews yet)")

4.3
3.9
4.1
3.8
3.9
1.8
3.7
4.3
index=  8 , 0 stars (no reviews yet)
4.1
5
4.3
index=  12 , 0 stars (no reviews yet)
4.6
3.6


In [614]:
print(soups[10].select('li[id^="result_"]')[0].find_all("a")[2]["href"]) # 5stars (although only 2 reviews)

https://www.amazon.com/gp/offer-listing/B01HQK8IZA/ref=lp_510114_1_241_olp/157-4244468-6130167?s=vacuums&ie=UTF8&qid=1485361987&sr=1-241&condition=new


In [615]:
print(soups[12].select('li[id^="result_"]')[0].find_all("a")[2]["href"]) # 0 start (no customer reviews yet)

https://www.amazon.com/BLACK-DECKER-DustBuster-Cordless-Vacuum/dp/B00NCT8F0S/ref=lp_510114_1_289/157-4916605-7228009?s=vacuums&ie=UTF8&qid=1485361993&sr=1-289


### Now we are ready to merge all the ingredients learned from above code blocks into one function

In [658]:
def items_info_extractor(soups):
    
    item_links=[]
    item_num_of_reviews=[]
    item_prices=[]
    item_names=[]
    item_ids=[]
    item_brands=[]
    item_avestars=[]
    
    for soup in soups:
        items=soup.select('li[id^="result_"]')

        for item in items:

            link_item=item.select("a[href$='customerReviews']")

            # ignore those items which contains 0 customer reviews. Those items are irrelevent to us.
            if (link_item !=[]):  

                price_tag=link_item[0].parent.previous_sibling.previous_sibling
                price_main_tag=price_tag.select(".sx-price-whole")
                price_fraction_tag=price_tag.select(".sx-price-fractional")

                link=link_item[0]["href"]

                # Ignore items which don't have normal price tags.
                # Those are items which are not sold by Amazon directly.
                # Also, remove those items which are ads (3 ads are shown in each page).
                if((price_main_tag !=[]) & (price_fraction_tag !=[]) & (link.endswith("spons#customerReviews") == False)):

                    # extract the item's name and ID from the obtained link
                    item_name=link.split("/")[3]
                    item_id=link.split("/")[5]
                    # replace the obtained link by the link that will lead to the customer reviews
                    base_url="https://www.amazon.com/"
                    link=base_url+item_name+"/product-reviews/"+item_id+"/ref=cm_cr_getr_d_paging_btm_" \
                                 +str(1)+"?ie=UTF8&pageNumber="+str(1)+"&reviewerType=all_reviews&pageSize=1000"

                    # obtain the price of the selected single item
                    price_main=price_main_tag[0].text
                    price_fraction=price_fraction_tag[0].text
                    item_price=int(price_main)+0.01*int(price_fraction)

                    # obtain the brand of the selected single item
                    item_brand=price_tag.parent.select(".a-size-small")[1].text
                    if(item_brand=="by "):
                        item_brand=price_tag.parent.select(".a-size-small")[2].text
                    # obtain the number of reviews of the selected single item
                    item_num_of_review=int(re.sub(",","",link_item[0].text))
                    
                    # obtain the averaged number of stars
                    starSelect=item.select_one("span[class='a-declarative']")
                    if((starSelect is None) or (starSelect.span is None)):  # there are no reviews yet (hence, we see no stars at all)
                        item_avestar=0
                    else:
                        item_avestar=starSelect.span.string.split(" ")[0]   # there are some reviews. So, we are able to extract the averaged number of stars
                    
                    # store the obtained variables into lists
                    item_links.append(link)
                    item_num_of_reviews.append(item_num_of_review)
                    item_prices.append(item_price)
                    item_names.append(item_name)
                    item_ids.append(item_id)
                    item_brands.append(item_brand)
                    item_avestars.append(item_avestar)
    return item_brands,item_ids,item_names,item_prices,item_num_of_reviews,item_links,item_avestars

In [659]:
item_brands,item_ids,item_names,item_prices,item_num_of_reviews,item_links,item_avestars=items_info_extractor(soups)

In [385]:
print(len(item_ids))
print(len(set(item_ids)))

387
387


In [386]:
print(len(item_names))
print(len(set(item_names)))

387
380


In [387]:
print(len(item_links))
print(len(set(item_links)))

387
387


The above results indicate that there are items that have the same product name but different links.

Cool. Let's find those products.

In [391]:
import collections
item_names_repeated=[]
for key in collections.Counter(item_names):
    if collections.Counter(item_names)[key]>1:
        print(key,collections.Counter(item_names)[key])
        item_names_repeated.append(key)
#print [item for item, count in collections.Counter(a).items() if count > 1]

ILIFE-Robotic-Cleaner-upgraded-Cleaning 2
Decker-Replacement-PHV1800-18-Volt-Pivoting 2
Shark-18V-Hand-Cordless-Vacuum 2
EcoGecko-Portable-Handheld-Mattress-Allergens 2
CHV1510-Dustbuster-15-6-Volt-Cordless-Cyclonic 2
Wrapables-Animal-Mini-Tabletop-Vacuum 2
Decker-9-6-Volt-Cordless-Dustbuster-BDH9600CHV 2


In [392]:
print(item_names_repeated)

['ILIFE-Robotic-Cleaner-upgraded-Cleaning', 'Decker-Replacement-PHV1800-18-Volt-Pivoting', 'Shark-18V-Hand-Cordless-Vacuum', 'EcoGecko-Portable-Handheld-Mattress-Allergens', 'CHV1510-Dustbuster-15-6-Volt-Cordless-Cyclonic', 'Wrapables-Animal-Mini-Tabletop-Vacuum', 'Decker-9-6-Volt-Cordless-Dustbuster-BDH9600CHV']


In [419]:
items_repeated=[]
for name,link,price,numrev in zip(item_names,item_links,item_prices,item_num_of_reviews):
    if name in item_names_repeated:
        #print(name,link,"\n")
        items_repeated.append((name,link,price,numrev))

sort a list with the method: sorted ( a "key" has to be given )

In [420]:
items_repeated=sorted(items_repeated, key=lambda x: x[0])

In [424]:
print("item name, item link, item price, total # of reviews of that item","\n")

for idx,(name,link,price,numrev) in enumerate(items_repeated):
    if((idx+1)%2==0):
        print(name,link,price,numrev,"\n")
    else:
        print(name,link,price,numrev)

item name, item link, item price, total # of reviews of that item 

CHV1510-Dustbuster-15-6-Volt-Cordless-Cyclonic https://www.amazon.com/CHV1510-Dustbuster-15-6-Volt-Cordless-Cyclonic/product-reviews/B00KASUEK8/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000 32.99 4793
CHV1510-Dustbuster-15-6-Volt-Cordless-Cyclonic https://www.amazon.com/CHV1510-Dustbuster-15-6-Volt-Cordless-Cyclonic/product-reviews/B004412GTO/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000 53.46 5088 

Decker-9-6-Volt-Cordless-Dustbuster-BDH9600CHV https://www.amazon.com/Decker-9-6-Volt-Cordless-Dustbuster-BDH9600CHV/product-reviews/B01JYUUQD2/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000 44.24 4
Decker-9-6-Volt-Cordless-Dustbuster-BDH9600CHV https://www.amazon.com/Decker-9-6-Volt-Cordless-Dustbuster-BDH9600CHV/product-reviews/B016P851MW/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&rev

What's found
* Each of the 7 items above has two different links/IDs (probably due to different color or seller) and varying prices.

## Now, let's try to merge the obtained data into pandas dataframe

Reference: http://pbpython.com/pandas-list-dict.html

In [661]:
for id in item_ids:
    if("B006LXOJC0" in id):
        print(id)

B006LXOJC0


In [664]:
df=pd.DataFrame.from_items([("pindex",item_ids),("type","handheld"),("pname",item_names),("brand",item_brands),("price",item_prices),("rurl",item_links),("totalRev",item_num_of_reviews),("avgStars",item_avestars)])

In [671]:
df.loc[:,["rurl","avgStars","totalRev"]]

rurl  \
0                        https://www.amazon.com/BLACK-DECKER-CHV1410L-Cordless-Lithium/product-reviews/B006LXOJC0/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
1                       https://www.amazon.com/Decker-BDH2000PL-Lithium-Vacuum-20-volt/product-reviews/B00IOEFBKS/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
2                https://www.amazon.com/Decker-HHVI320JR02-Dustbuster-Cordless-Lithium/product-reviews/B01DAI5CF6/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
3                        https://www.amazon.com/Black-Decker-HNV215B10-Compact-Lithium/product-reviews/B01BXBX6E6/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
4                       https://www.amazon.com/Dirt-Devil-SD20005RED-Scorpion-Handheld/product-reviews/B002D47XOM/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
5                            https://www.amazon.com/Dyson-V8-Absolute-Cord-Free-Vacuum/product-reviews/B01IENFJ14/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
6                             https://www.amazon.com/Dyson-Motor-Head-Cord-free-Vacuum/product-reviews/B00SMLJQ72/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
7                              https://www.amazon.com/Dyson-V6-Animal-Cord-free-Vacuum/product-reviews/B00SMLJQ7W/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
8                           https://www.amazon.com/Bissell-Eraser-Handheld-Vacuum-33A1/product-reviews/B001EYFQ28/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
9                         https://www.amazon.com/Eureka-EasyClean-Corded-Hand-Held-71B/product-reviews/B0006HUYGM/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
10                            https://www.amazon.com/Shark-Cordless-Perfect-Hand-SV780/product-reviews/B0037HHFMO/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
11               https://www.amazon.com/CHV1510-Dustbuster-15-6-Volt-Cordless-Cyclonic/product-reviews/B00KASUEK8/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
12                             https://www.amazon.com/Hair-Eraser-Cordless-Hand-Vacuum/product-reviews/B01E0472TI/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
13                   https://www.amazon.com/Handheld-Cordless-Cleaner-Lithium-Cyclonic/product-reviews/B01MDT3U9N/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
14                  https://www.amazon.com/Decker-HNV220BCZ01FF-Compact-Lithium-Vacuum/product-reviews/B01BXBX6CI/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
15                               https://www.amazon.com/Shark-Rocket-Corded-Hand-HV292/product-reviews/B00P9Z36T8/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
16                    https://www.amazon.com/BDH2000L-20-Volt-Lithium-Battery-Cordless/product-reviews/B008R3F0J8/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
17                                    https://www.amazon.com/Dyson-V6-Cord-Free-Vacuum/product-reviews/B00SMLJPIC/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
18                  https://www.amazon.com/Dyson-Animal-Cordless-Certified-Refurbished/product-reviews/B01AVXFD6Q/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   
19                     https://www.amazon.com/BLACK-DECKER-

#### Let's upload the obtained dataframe to MariaDB

In [466]:
from sqlalchemy import create_engine,Table,Column,Integer,String,MetaData,ForeignKey,Date
import pymysql

engine=create_engine("mysql+pymysql://semantic:GbwSq1RzFb@104.199.201.206:13606/Tests?charset=utf8",echo=False, encoding='utf-8')
conn = engine.connect()

df.to_sql(name='amzProd', con=conn, if_exists = 'append', index=False)
conn.close()

/usr/local/lib/python3.5/site-packages/pymysql/cursors.py:166: Warning: (1265, "Data truncated for column 'pname' at row 80")
  result = self._query(query)
/usr/local/lib/python3.5/site-packages/pymysql/cursors.py:166: Warning: (1265, "Data truncated for column 'brand' at row 313")
  result = self._query(query)


#### Alternatively, we can store the obtained dataframe into a csv file

In [469]:
df.to_csv("ProdInfo_handheld_26012017.csv", encoding="utf-8")

And load it:

In [474]:
pd.DataFrame.from_csv("ProdInfo_handheld_26012017.csv", encoding="utf-8")


pindex      type                                              pname  \
0    B006LXOJC0  handheld             BLACK-DECKER-CHV1410L-Cordless-Lithium   
1    B00IOEFBKS  handheld            Decker-BDH2000PL-Lithium-Vacuum-20-volt   
2    B01DAI5CF6  handheld     Decker-HHVI320JR02-Dustbuster-Cordless-Lithium   
3    B01BXBX6E6  handheld             Black-Decker-HNV215B10-Compact-Lithium   
4    B002D47XOM  handheld            Dirt-Devil-SD20005RED-Scorpion-Handheld   
5    B01IENFJ14  handheld                 Dyson-V8-Absolute-Cord-Free-Vacuum   
6    B00SMLJQ72  handheld                  Dyson-Motor-Head-Cord-free-Vacuum   
7    B00SMLJQ7W  handheld                   Dyson-V6-Animal-Cord-free-Vacuum   
8    B001EYFQ28  handheld                Bissell-Eraser-Handheld-Vacuum-33A1   
9    B0006HUYGM  handheld              Eureka-EasyClean-Corded-Hand-Held-71B   
10   B0037HHFMO  handheld                  Shark-Cordless-Perfect-Hand-SV780   
11   B00KASUEK8  handheld     CHV1510-Dustbuster-15-6-Volt-Cordless-Cyclonic   
12   B01E0472TI  handheld                   Hair-Eraser-Cordless-Hand-Vacuum   
13   B01MDT3U9N  handheld         Handheld-Cordless-Cleaner-Lithium-Cyclonic   
14   B01BXBX6CI  handheld        Decker-HNV220BCZ01FF-Compact-Lithium-Vacuum   
15   B00P9Z36T8  handheld                     Shark-Rocket-Corded-Hand-HV292   
16   B008R3F0J8  handheld          BDH2000L-20-Volt-Lithium-Battery-Cordless   
17   B00SMLJPIC  handheld                          Dyson-V6-Cord-Free-Vacuum   
18   B01AVXFD6Q  handheld        Dyson-Animal-Cordless-Certified-Refurbished   
19   B00JILGZOC  handheld           BLACK-DECKER-BDH2020FLFH-Lithium-20-volt   
20   B00IRJ1BYU  handheld             BISSELL-SPOTLIFTER-ESSENTIAL-RED-1719T   
21   B000050HCV  handheld              Dirt-Devil-Cleaner-Handheld-M08230RED   
22   B00ECM5RRI  handheld        Black-Decker-BDH9600CHV-Dustbuster-Cordless   
23   B001AQEQNA  handheld  PAV1200W-Cyclonic-Action-Automotive-Pivoting-N...   
24   B0007Z69G2  handheld             Bissell-Aeroswift-Compact-Bagless-1009   
25   B002FQJW4W  handheld          Stanley-Decker-FHV1200W-Cordless-Canister   
26   B01DAI5BZ2  handheld     Decker-HHVI315JO42-Dustbuster-Cordless-Lithium   
27   B00SWCU1LG  handheld         Hoover-Cleaner-Cordless-Handheld-BH52120PC   
28   B00CCYLBZ0  handheld               BLACK-DECKER-BDH2020FL-Lithium-Brush   
29   B00489LHEM  handheld              Dirt-Devil-BD10025WX-Bagless-Cordless   
..          ...       ...                                                ...   
357  B008JGXHZW  handheld           CleanWave-UV-C-Vacuum-Cleaner-Prevention   
358  B00440E50I  handheld                       Bionaire-Hand-Held-Turbo-Vac   
359  B0019L6YU4  handheld                  HAMMER-Dirt-Devil-Scorpion-Filter   
360  B00JWM9CHO  handheld         Creative-Fashion-Keyboard-Desktop-Handheld   
361  B01LVXEM9Y  handheld               Vacuum-Company-Car-Cleaner-CV-LDA105   
362  B00DOTD7KY  handheld                  Kole-Imports-CA026-Vacuum-Cleaner   
363  B000UKQMLG  handheld   Honeywell-H19003-Replacement-DustBuster-Cyclonic   
364  B000FMIRLM  handheld                      WackyVac-Wacky-Vac-Attachment   
365  B00JEMTJ66  handheld                 Metrovac-Electra-Sweep-Power-Broom   
366  B00GQ4LH06  handheld     Iautomatic-Cleaner-Accessory-suction-1-25-Inch   
367  B00CDTPGVO  handheld                Koblenz-HV-120-Corded-120-Volt-Hand   
368  B008CCQJUI  handheld   Unilution-75210-White-Portable-Handheld-Mattress   
369  B007DIT1KW  handheld         DECKER-CHV4800-BUSTER-CORDLESS-KITCHENWARE   
370  B005HP495A  handheld                   AUVWD2-Dirt-Magic-Vacuum-Cleaner   
371  B004H6PBUM  handheld      EcoGecko-Portable-Handheld-Mattress-Allergens   
372  B001SN8FAA  handheld      Panasonic-handheld-vacuum-cleaner-MC-D25CP-WA   
373  B000SMOFHE  handheld    Royal-Appliance-3-089570-001-Replacement-Filter   
374  B01MQE6O7Y  handheld                Atrix-VACAMEXP-Express-Handi-Vacuum   
375  B01F

#### Upload the obtained CSV files to the remote MariaDB

In [496]:
from sqlalchemy import create_engine,Table,Column,Integer,String,MetaData,ForeignKey,Date
import pymysql
import datetime

I found out that there might be same pindex in one dataframe. This can lead to an error if we are going to upload our data to MariaDB, as the primary key is ought to be unique.

In [531]:
pd.set_option('max_colwidth', 800)
for idx,df in enumerate(dfs):
    print(idx,df.loc[df['pindex'] == 'B00SWGVICS'])

0 Empty DataFrame
Columns: [pindex, type, pname, brand, price, rurl, totalRev]
Index: []
1         pindex      type                                      pname  \
34  B00SWGVICS  canister  Hoover-Commercial-CH32008-Canister-Vacuum   

                brand  price  \
34  Hoover Commercial  141.0   

                                                                                                                                                                                     rurl  \
34  https://www.amazon.com/Hoover-Commercial-CH32008-Canister-Vacuum/product-reviews/B00SWGVICS/ref=cm_cr_getr_d_paging_btm_1?ie=UTF8&pageNumber=1&reviewerType=all_reviews&pageSize=1000   

    totalRev  
34        54  
2         pindex      type                                      pname  \
68  B00SWGVICS  handheld  Hoover-Commercial-CH32008-Canister-Vacuum   

                brand  price  \
68  Hoover Commercial  141.0   

                                                                                  

## Strategy: Store all csvs into one dataframe. Then, remove all duplicates before uploading to the DataBase.

In [42]:
import os
from IPython.display import display

In [32]:
cwd=os.getcwd()

In [33]:
print(cwd)

/Users/chweng/Google Drive/SemanticProj/webCrawler


----

Now, it's time to get to know the Pandas Dataframe better. I'd like to figure out how two dataframes can be merged horizontally.

#### an one column example: pd.Dataframe.from_items()

In [58]:
test_col = pd.DataFrame.from_items([("test_column1",np.arange(10))])
test_col2 = pd.DataFrame.from_items([("test_column2",5+np.arange(10))])
display(test_col,test_col2)

test_column1
0             0
1             1
2             2
3             3
4             4
5             5
6             6
7             7
8             8
9             9

test_column2
0             5
1             6
2             7
3             8
4             9
5            10
6            11
7            12
8            13
9            14

In [59]:
result = pd.concat([test_col, test_col2], axis=1)

In [60]:
display(result)

test_column1  test_column2
0             0             5
1             1             6
2             2             7
3             3             8
4             4             9
5             5            10
6             6            11
7             7            12
8             8            13
9             9            14

---

In [112]:
date="2017-01-30"
prodTypes=["central","canister","handheld","robotic","stick","upright","wetdry"]

# put all the dataframes into a list
dfs=[pd.DataFrame.from_csv("data/ProdInfo_%s_%s.csv"%(prodType,date), encoding="utf-8") for prodType in prodTypes]


for idx,df in enumerate(dfs):
    cID=[j%7 for j in range(df.shape[0])]
    colCID=pd.DataFrame.from_items([( "cID",cID )])
    dfs[idx]=pd.concat([df, colCID], axis=1)

# concatenate dataframes
df=pd.concat(dfs).drop_duplicates("rurl")

# prepare the connection and connect to the DB
engine=create_engine("mysql+pymysql://semantic:GbwSq1RzFb@104.199.201.206:13606/Tests?charset=utf8",echo=False, encoding='utf-8')
conn = engine.connect()

# remove duplicates and upload the concatenated dataframe to the SQL DataBase
df.to_sql(name='amzProd', con=conn, if_exists = 'append', index=False)

# close the connection
conn.close()

In [111]:
len(df.iloc[974]["brand"])

60

In [540]:
df.iloc[463]["pname"]

'Handheld-Vacuum-Cleaner-Abask-Vacuum-Cleaner-7-2V-60W-Ni-CD2200MA-3-5KPA-Suction-Portable-1-Accessories-Rechargeable-Cordless-Cleaner'

In [543]:
!echo "Handheld-Vacuum-Cleaner-Abask-Vacuum-Cleaner-7-2V-60W-Ni-CD2200MA-3-5KPA-Suction-Portable-1-Accessories-Rechargeable-Cordless-Cleaner"| wc 

       1       1     134


Length of this string is larger than 100. Therefore, I have to alter our schema, since the product name was set to have length 100 by default.